lib/apiservers/engine/backends/container.go

```diff
@@ -1605,6 +1600,70 @@ func (c *Container) containerAttach(name string, ca *backend.ContainerAttachConf
 	return nil
 }
 
+// ContainerRename changes the name of a container, using the oldName
+// to find the container. An error is returned if newName is already
+// reserved.
+func (c *Container) ContainerRename(oldName, newName string) error {
+	defer trace.End(trace.Begin(newName))
+
+	if oldName == "" || newName == "" {
+		err := fmt.Errorf("neither old nor new names may be empty")
+		log.Errorf("%s", err.Error())
+		return derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	if !utils.RestrictedNamePattern.MatchString(newName) {
+		err := fmt.Errorf("invalid container name (%s), only %s are allowed", newName, utils.RestrictedNameChars)
+		log.Errorf("%s", err.Error())
+		return derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(oldName)
+	if vc == nil {
+		log.Errorf("Container %s not found", oldName)
+		return NotFoundError(oldName)
+	}
+
+	oldName = vc.Name
+	if oldName == newName {
+		err := fmt.Errorf("renaming a container with the same name as its current name")
+		log.Errorf("%s", err.Error())
+		return derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	if exists := cache.ContainerCache().GetContainer(newName); exists != nil {
```

Comment:
@fdawg4l ReserveName only checks the map `containersByName` in containerCache and then reserve that name in the cache. However, `GetContainer` will check all the maps in the cache, although it is really rare that the `newName` could be the same with the `id` of an existing container.